File: heap-vs-stack/stack_vs_heap.cpp
#include <iostream>
#include <chrono>

class LargeObject {
public:
    int data[1000];
    LargeObject() { data[0] = 42; }
};

void stackAllocation() {
    auto start = std::chrono::high_resolution_clock::now();
    LargeObject obj;
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start);
    std::cout << "Stack allocation time: " << duration.count() << " ns" << std::endl;
    std::cout << "Object data: " << obj.data[0] << std::endl;
}

void heapAllocation() {
    auto start = std::chrono::high_resolution_clock::now();
    LargeObject* obj = new LargeObject();
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start);
    std::cout << "Heap allocation time: " << duration.count() << " ns" << std::endl;
    std::cout << "Object data: " << obj->data[0] << std::endl;
    delete obj;
}

int main() {
    std::cout << "=== Stack vs Heap Comparison ===" << std::endl;
    std::cout << "\nKey Differences:" << std::endl;
    std::cout << "1. Stack: Fast, automatic cleanup, limited size" << std::endl;
    std::cout << "2. Heap: Slower, manual cleanup, large size" << std::endl;
    std::cout << "\nPerformance comparison:\n" << std::endl;
    stackAllocation();
    heapAllocation();
    return 0;
}

File: heap-vs-stack/stack_vs_heap.rs
use std::time::Instant;

struct LargeObject {
    data: [i32; 1000],
}

impl LargeObject {
    fn new() -> Self {
        LargeObject { data: [0; 1000] }
    }
}

fn stack_allocation() {
    let start = Instant::now();
    let obj = LargeObject::new();
    let duration = start.elapsed();
    println!("Stack allocation time: {:?}", duration);
    println!("Object data: {}", obj.data[0]);
}

fn heap_allocation() {
    let start = Instant::now();
    let obj = Box::new(LargeObject::new());
    let duration = start.elapsed();
    println!("Heap allocation time: {:?}", duration);
    println!("Object data: {}", obj.data[0]);
}

fn demonstrate_ownership() {
    println!("\n=== Rust Ownership Demo ===");
    let stack_str = "Hello, Stack!";
    println!("Stack string: {}", stack_str);
    let heap_str = String::from("Hello, Heap!");
    println!("Heap string: {}", heap_str);
    let moved_str = heap_str;
    println!("Moved string: {}", moved_str);
}

fn main() {
    println!("=== Stack vs Heap Comparison ===");
    println!("\nKey Differences:");
    println!("1. Stack: Fast, automatic cleanup, limited size, Copy semantics");
    println!("2. Heap: Slower, automatic cleanup (RAII), large size, Move semantics");
    println!("\nPerformance comparison:\n");
    stack_allocation();
    heap_allocation();
    demonstrate_ownership();
}
